pkg/credentials/credentials.go

```diff
@@ -283,13 +287,45 @@ func defaultValidationFunc(creds map[string]string) error {
 }
 
 func openstackValidationFunc(creds map[string]string) error {
-	for k, v := range creds {
-		if k == OpenStackTenantID || k == OpenStackTenantName {
-			continue
+	alwaysRequired := []string{OpenStackAuthURL, OpenStackDomainName, OpenStackRegionName}
+
+	for _, key := range alwaysRequired {
+		if v, ok := creds[key]; !ok || len(v) == 0 {
+			return errors.Errorf("key %v is required but is not present", key)
 		}
-		if len(v) == 0 {
-			return errors.Errorf("key %v is required but isn't present", k)
+	}
+
+	var (
+		appCredsInUse  bool
+		userCredsInUse bool
+	)
+
+	if v, ok := creds[OpenStackApplicationCredentialID]; ok && len(v) != 0 {
+		if v, ok := creds[OpenStackApplicationCredentialSecret]; !ok || len(v) == 0 {
+			return errors.Errorf("key %v is required if %v is present", OpenStackApplicationCredentialSecret, OpenStackApplicationCredentialID)
 		}
+		appCredsInUse = true
+	}
+
+	if v, ok := creds[OpenStackUserName]; !appCredsInUse && ok && len(v) != 0 {
+		if v, ok := creds[OpenStackPassword]; !ok || len(v) == 0 {
+			return errors.Errorf("key %v is required but is not present", OpenStackPassword)
+		}
+		userCredsInUse = true
+	}
+
+	if !appCredsInUse && !userCredsInUse {
+		return errors.Errorf("no app credentials (%s, %s) or user credentials (%s, %s) found",
+			OpenStackApplicationCredentialID, OpenStackApplicationCredentialSecret,
+			OpenStackUserName, OpenStackPassword,
+		)
+	}
+
+	if appCredsInUse && userCredsInUse {
+		return errors.Errorf("both app credentials (%s, %s) and user credentials (%s, %s) defined",
+			OpenStackApplicationCredentialID, OpenStackApplicationCredentialSecret,
+			OpenStackUserName, OpenStackPassword,
+		)
 	}
```

Comment:
Should we care about this at all? The GopherCloud library used by Terraform and machine-controller knows how to handle this case -- [by preferring the Application Credentials](https://github.com/gophercloud/gophercloud/blob/master/openstack/auth_env.go).

Even if both are provided, we would just put them in the Secret, and the library would figure out what to use. However, a good thing in this approach is that we are explicit -- we are not going to put something in the Secret that we are not going to use, therefore, we're improving the security.

cc: @kron4eg (for opinion)

Reply:
Subjectively, I like this. It might be an error from the user that they have both in their ENV. We might accidentally leak non-application credentials, so having such error is good.

Reply:
Yeah, that's what I thought as well -- only one type of credentials can be used at the time anyways. Even if they have both, they can always temporarily `unset` credentials they don't want to use/deploy.